File: solutions/prolog/nth-prime/1/nth_prime.pl
% Dedicated to Shree DR.MDD

is_prime(2) :- !.
is_prime(3) :- !.
is_prime(X) :- X > 3, X mod 2 =\= 0, is_prime(X, 3), !.
is_prime(X, D) :- (D*D > X -> true; X mod D =\= 0, D1 is D + 2, is_prime(X, D1)), !.

find_next(P, P) :- is_prime(P), !.
find_next(P, R) :- P1 is P + 1, find_next(P1, R).

nth_prime(1, 2) :- !.
nth_prime(2, 3) :- !.
nth_prime(K, R) :-
    K > 2,
    K1 is K - 1,
    nth_prime(K1, Prev),
    S is Prev + 1,
    find_next(S, R).

File: solutions/fortran/nth-prime/1/nth_prime.f90
! Dedicated to Shree DR.MDD  
module nth_prime
   implicit none
contains

   integer function prime(n)
      integer, intent(in) :: n
      integer, dimension(n) :: primeList
      integer :: pos
      integer :: cand, k
      logical :: isFound

      pos = 1
      isFound = .false.

      if (n < 1) then
         prime = -1
         return
      end if

      primeList(1) = 2

      do while (pos < n)

         cand = primeList(pos) + 1
         isFound = .false.

         search_loop: do while (.not. isFound)
            do k = 1, pos
               if (mod(cand, primeList(k)) == 0) then
                  cand = cand + 1
                  cycle search_loop
               end if
            end do
            pos = pos + 1
            primeList(pos) = cand
            isFound = .true.
         end do search_loop

      end do

      prime = primeList(n)

   end function

end module nth_prime
